File: labs/19-long-task-reliability/go/main.go
package main

import (
	"crypto/sha1"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"log"
	"math"
	"math/rand"
	"os"
	"path/filepath"
	"strings"
	"time"

	openaiclient "github.com/ZYZ-Labs/agent-labs/shared/config"
)

var checkpointDir = ".checkpoints"

type checkpointStore struct {
	path string
}

func newCheckpointStore(key string) *checkpointStore {
	return &checkpointStore{path: filepath.Join(checkpointDir, key+".json")}
}

func (s *checkpointStore) load() map[string]interface{} {
	b, err := os.ReadFile(s.path)
	if err != nil {
		return map[string]interface{}{}
	}
	var state map[string]interface{}
	_ = json.Unmarshal(b, &state)
	return state
}

func (s *checkpointStore) save(state map[string]interface{}) error {
	if err := os.MkdirAll(checkpointDir, 0755); err != nil {
		return err
	}
	b, err := json.MarshalIndent(state, "", "  ")
	if err != nil {
		return err
	}
	tmp := s.path + ".tmp"
	if err := os.WriteFile(tmp, b, 0644); err != nil {
		return err
	}
	return os.Rename(tmp, s.path)
}

func exponentialBackoff(attempt int, baseDelay, maxDelay float64) time.Duration {
	delay := math.Min(baseDelay*math.Pow(2, float64(attempt-1)), maxDelay)
	jitter := delay * 0.1
	if attempt%2 == 0 {
		jitter = -jitter
	}
	return time.Duration((delay+jitter)*1000) * time.Millisecond
}

func runWithTimeout(fn func() (interface{}, error), timeout time.Duration) (interface{}, error) {
	type result struct {
		val interface{}
		err error
	}
	ch := make(chan result, 1)
	go func() {
		val, err := fn()
		ch <- result{val, err}
	}()
	select {
	case r := <-ch:
		return r.val, r.err
	case <-time.After(timeout):
		return nil, fmt.Errorf("step timed out after %v", timeout)
	}
}

type longTask struct {
	inputData       string
	client          *openaiclient.Client
	idempotencyKey  string
	store           *checkpointStore
	state           map[string]interface{}
	flakyAttempts   int
}

func newLongTask(inputData string, client *openaiclient.Client) *longTask {
	key := makeKey(inputData)
	store := newCheckpointStore(key)
	state := store.load()
	return &longTask{
		inputData:      inputData,
		client:         client,
		idempotencyKey: key,
		store:          store,
		state:          state,
	}
}

func makeKey(inputData string) string {
	h := sha1.New()
	_, _ = h.Write([]byte(inputData))
	return "task-" + hex.EncodeToString(h.Sum(nil))
}

func (t *longTask) isComplete() bool {
	status, _ := t.state["status"].(string)
	key, _ := t.state["idempotency_key"].(string)
	return status == "completed" && key == t.idempotencyKey
}

func (t *longTask) runStep(name string, fn func() (interface{}, error), timeout time.Duration, maxRetries int) (interface{}, error) {
	completedSteps, _ := t.state["completed_steps"].(map[string]interface{})
	if completedSteps == nil {
		completedSteps = map[string]interface{}{}
	}
	if completedSteps[name] == true {
		log.Printf("Step '%s' already completed; skipping", name)
		results, _ := t.state["results"].(map[string]interface{})
		return results[name], nil
	}

	log.Printf("Executing step '%s'", name)
	var lastErr error
	for attempt := 1; attempt <= maxRetries; attempt++ {
		result, err := runWithTimeout(fn, timeout)
		if err == nil {
			if completedSteps == nil {
				completedSteps = map[string]interface{}{}
			}
			completedSteps[name] = true
			t.state["completed_steps"] = completedSteps
			results, _ := t.state["results"].(map[string]interface{})
			if results == nil {
				results = map[string]interface{}{}
			}
			results[name] = result
			t.state["results"] = results
			t.state["last_step"] = name
			if err := t.store.save(t.state); err != nil {
				log.Printf("Checkpoint save failed: %v", err)
			}
			log.Printf("Step '%s' succeeded", name)
			return result, nil
		}
		lastErr = err
		log.Printf("Step '%s' attempt %d failed: %v", name, attempt, err)
		if attempt < maxRetries {
			delay := exponentialBackoff(attempt, 1.0, 30.0)
			log.Printf("Retrying step '%s' in %.2fs", name, delay.Seconds())
			time.Sleep(delay)
		}
	}
	log.Printf("Step '%s' exhausted retries", name)
	return nil, lastErr
}

func (t *longTask) stepFetchData() (interface{}, error) {
	log.Printf("Fetching data for: %s", t.inputData)
	if t.client != nil {
		resp, err := t.client.ChatCompletion(openaiclient.ChatCompletionRequest{
			Messages:  []openaiclient.Message{{Role: "user", Content: fmt.Sprintf("Summarize '%s' in one sentence.", t.inputData)}},
			MaxTokens: intPtr(50),
		})
		if err != nil {
			return nil, err
		}
		return strings.TrimSpace(resp.Choices[0].Message.Content), nil
	}
	return fmt.Sprintf("Mock summary for '%s'.", t.inputData), nil
}

func (t *longTask) stepProcessData(fetched string) (interface{}, error) {
	t.flakyAttempts++
	if t.flakyAttempts < 3 {
		return nil, fmt.Errorf("processing service busy (attempt %d)", t.flakyAttempts)
	}
	return fmt.Sprintf("processed(%s)", fetched), nil
}

func (t *longTask) stepNotify(processed string) (interface{}, error) {
	return fmt.Sprintf("notification_sent(%s)", processed), nil
}

func (t *longTask) run() (map[string]interface{}, error) {
	if t.isComplete() {
		log.Printf("Task already completed for key %s; returning cached result", t.idempotencyKey)
		results, _ := t.state["results"].(map[string]interface{})
		return map[string]interface{}{"status": "completed", "results": results}, nil
	}

	t.state["idempotency_key"] = t.idempotencyKey
	t.state["status"] = "running"
	if _, ok := t.state["completed_steps"]; !ok {
		t.state["completed_steps"] = map[string]interface{}{}
	}
	if _, ok := t.state["results"]; !ok {
		t.state["results"] = map[string]interface{}{}
	}

	fetched, err := t.runStep("fetch", t.stepFetchData, 5*time.Second, 3)
	if err != nil {
		return nil, err
	}
	processed, err := t.runStep("process", func() (interface{}, error) { return t.stepProcessData(fetched.(string)) }, 5*time.Second, 3)
	if err != nil {
		return nil, err
	}
	notified, err := t.runStep("notify", func() (interface{}, error) { return t.stepNotify(processed.(string)) }, 5*time.Second, 3)
	if err != nil {
		return nil, err
	}
	_ = notified

	t.state["status"] = "completed"
	if err := t.store.save(t.state); err != nil {
		log.Printf("Final checkpoint save failed: %v", err)
	}
	log.Println("Task completed successfully")
	results, _ := t.state["results"].(map[string]interface{})
	return map[string]interface{}{"status": "completed", "results": results}, nil
}

func intPtr(i int) *int {
	return &i
}

func init() {
	rand.Seed(time.Now().UnixNano())
}

func main() {
	cfg := openaiclient.LoadConfig()
	var client *openaiclient.Client
	if cfg.APIKey != "" || strings.HasPrefix(cfg.BaseURL, "http://localhost") {
		client = openaiclient.NewClient()
	} else {
		log.Println("LLM client disabled: OPENAI_API_KEY not set")
	}

	inputData := "reliable agent engineering"
	task := newLongTask(inputData, client)

	fmt.Println("Starting long-running task...")
	fmt.Println("Idempotency key:", task.idempotencyKey)
	result, err := task.run()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Task failed: %v\n", err)
		os.Exit(1)
	}
	fmt.Println("\nFinal result:")
	b, _ := json.MarshalIndent(result, "", "  ")
	fmt.Println(string(b))

	fmt.Println("\nRe-running with the same idempotency key...")
	task2 := newLongTask(inputData, client)
	result2, err := task2.run()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Task failed: %v\n", err)
		os.Exit(1)
	}
	b, _ = json.MarshalIndent(result2, "", "  ")
	fmt.Println(string(b))
}

File: labs/19-long-task-reliability/node/main.ts
import fs from "fs";
import path from "path";
import { createHash } from "crypto";
import { OpenAIClient } from "../../../shared/config/openai_client";

const CHECKPOINT_DIR = path.join(__dirname, ".checkpoints");
fs.mkdirSync(CHECKPOINT_DIR, { recursive: true });

class CheckpointStore {
  path: string;

  constructor(key: string) {
    this.path = path.join(CHECKPOINT_DIR, `${key}.json`);
  }

  load(): any {
    if (fs.existsSync(this.path)) {
      return JSON.parse(fs.readFileSync(this.path, "utf8"));
    }
    return {};
  }

  save(state: any): void {
    const tmp = `${this.path}.tmp`;
    fs.writeFileSync(tmp, JSON.stringify(state, null, 2), "utf8");
    fs.renameSync(tmp, this.path);
  }
}

function exponentialBackoff(attempt: number, baseDelay = 1.0, maxDelay = 30.0): number {
  const delay = Math.min(baseDelay * Math.pow(2, attempt - 1), maxDelay);
  const jitter = delay * 0.1 * (attempt % 2 === 0 ? 1 : -1);
  return delay + jitter;
}

function runWithTimeout<T>(func: () => T | Promise<T>, timeout: number): Promise<T> {
  return Promise.race([
    Promise.resolve(func()),
    new Promise<T>((_, reject) =>
      setTimeout(
        () => reject(new Error(`Step timed out after ${timeout}s`)),
        timeout * 1000
      )
    ),
  ]);
}

class LongTask {
  idempotencyKey: string;
  store: CheckpointStore;
  state: any;
  private flakyAttempts = 0;

  constructor(private inputData: string, private client: OpenAIClient | null) {
    this.idempotencyKey = this.makeKey(inputData);
    this.store = new CheckpointStore(this.idempotencyKey);
    this.state = this.store.load();
  }

  private makeKey(inputData: string): string {
    const key = createHash("sha1")
      .update("6ba7b810-9dad-11d1-80b4-00c04fd430c8" + inputData)
      .digest("hex");
    return `task-${key}`;
  }

  private isComplete(): boolean {
    return (
      this.state.status === "completed" &&
      this.state.idempotency_key === this.idempotencyKey
    );
  }

  private async runStep(
    name: string,
    func: () => any,
    timeout = 5.0,
    maxRetries = 3
  ): Promise<any> {
    if (this.state.completed_steps?.[name]) {
      console.log(`Step '${name}' already completed; skipping`);
      return this.state.results[name];
    }

    console.log(`Executing step '${name}'`);
    let lastError: Error | null = null;
    for (let attempt = 1; attempt <= maxRetries; attempt++) {
      try {
        const result = await runWithTimeout(func, timeout);
        if (!this.state.completed_steps) this.state.completed_steps = {};
        this.state.completed_steps[name] = true;
        if (!this.state.results) this.state.results = {};
        this.state.results[name] = result;
        this.state.last_step = name;
        this.store.save(this.state);
        console.log(`Step '${name}' succeeded`);
        return result;
      } catch (exc: any) {
        lastError = exc;
        console.warn(`Step '${name}' attempt ${attempt} failed: ${exc.message}`);
        if (attempt < maxRetries) {
          const delay = exponentialBackoff(attempt);
          console.log(`Retrying step '${name}' in ${delay.toFixed(2)}s`);
          await new Promise((r) => setTimeout(r, delay * 1000));
        }
      }
    }
    console.error(`Step '${name}' exhausted retries`);
    throw lastError;
  }

  private async stepFetchData(): Promise<string> {
    console.log(`Fetching data for: ${this.inputData}`);
    if (this.client) {
      const resp = await this.client.chatCompletion({
        messages: [
          { role: "user", content: `Summarize '${this.inputData}' in one sentence.` },
        ],
        max_tokens: 50,
      });
      return resp.choices[0].message.content.trim();
    }
    return `Mock summary for '${this.inputData}'.`;
  }

  private stepProcessData(fetched: string): string {
    this.flakyAttempts += 1;
    if (this.flakyAttempts < 3) {
      throw new Error(`Processing service busy (attempt ${this.flakyAttempts})`);
    }
    return `processed(${fetched})`;
  }

  private stepNotify(processed: string): string {
    return `notification_sent(${processed})`;
  }

  async run(): Promise<{ status: string; results: any }> {
    if (this.isComplete()) {
      console.log(`Task already completed for key ${this.idempotencyKey}; returning cached result`);
      return { status: "completed", results: this.state.results };
    }

    this.state.idempotency_key = this.idempotencyKey;
    this.state.status = this.state.status || "running";
    this.state.completed_steps = this.state.completed_steps || {};
    this.state.results = this.state.results || {};

    const fetched = await this.runStep("fetch", () => this.stepFetchData());
    const processed = await this.runStep("process", () => this.stepProcessData(fetched));
    const notified = await this.runStep("notify", () => this.stepNotify(processed));

    this.state.status = "completed";
    this.store.save(this.state);
    console.log("Task completed successfully");
    return { status: "completed", results: this.state.results };
  }
}

async function main() {
  let client: OpenAIClient | null = null;
  try {
    client = new OpenAIClient();
  } catch (exc: any) {
    console.warn("LLM client disabled:", exc.message);
  }

  const inputData = "reliable agent engineering";
  const task = new LongTask(inputData, client);

  console.log("Starting long-running task...");
  console.log("Idempotency key:", task.idempotencyKey);
  const result = await task.run();
  console.log("\nFinal result:");
  console.log(JSON.stringify(result, null, 2));

  console.log("\nRe-running with the same idempotency key...");
  const task2 = new LongTask(inputData, client);
  const result2 = await task2.run();
  console.log(JSON.stringify(result2, null, 2));
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
